tests/test_simulation.py

```diff
@@ -97,21 +98,18 @@ def testSimulationVolCoords(self):
 
     def testSimulationCovar(self):
         covar = self.sim.covar_true()
-        self.assertTrue(
-            np.allclose(
-                [
-                    [-0.00000289, -0.00005839, -0.00018998, -0.00124722, -0.00003155, +0.00743356, +0.00798143, +0.00303416],
-                    [-0.00000776, +0.00018371, +0.00448675, -0.00794970, -0.02988000, -0.00185446, +0.01786612, +0.00685990],
-                    [+0.00001144, +0.00324029, +0.03364052, -0.00272520, -0.08976389, -0.05404807, +0.00268740, -0.03081760],
-                    [+0.00003204, +0.00909853, +0.07859941, +0.07254293, -0.19365733, -0.09007251, -0.15731451, -0.15690306],
-                    [-0.00040561, +0.00685139, +0.11074986, +0.35207557, +0.17264650, -0.16662873, -0.15010859, -0.14292650],
-                    [-0.00107461, -0.00497393, +0.04630126, +0.38048555, +0.47915877, +0.05379957, -0.11833663, -0.03372971],
-                    [-0.00029630, -0.00485664, -0.00640120, +0.22068169, +0.15419035, +0.08281200, +0.03373241, +0.00103902],
-                    [+0.00044323, +0.00850533, +0.09683860, +0.16959519, +0.03629097, +0.03740599, +0.02212356, +0.00318127],
-                ],
-                covar[:, :, 4, 4, 4, 4]
-            )
-        )
+        result = [
+            [-0.00000289, -0.00005839, -0.00018998, -0.00124722, -0.00003155, +0.00743356, +0.00798143, +0.00303416],
+            [-0.00000776, +0.00018371, +0.00448675, -0.00794970, -0.02988000, -0.00185446, +0.01786612, +0.00685990],
+            [+0.00001144, +0.00324029, +0.03364052, -0.00272520, -0.08976389, -0.05404807, +0.00268740, -0.03081760],
+            [+0.00003204, +0.00909853, +0.07859941, +0.07254293, -0.19365733, -0.09007251, -0.15731451, -0.15690306],
+            [-0.00040561, +0.00685139, +0.11074986, +0.35207557, +0.17264650, -0.16662873, -0.15010859, -0.14292650],
+            [-0.00107461, -0.00497393, +0.04630126, +0.38048555, +0.47915877, +0.05379957, -0.11833663, -0.03372971],
+            [-0.00029630, -0.00485664, -0.00640120, +0.22068169, +0.15419035, +0.08281200, +0.03373241, +0.00103902],
+            [+0.00044323, +0.00850533, +0.09683860, +0.16959519, +0.03629097, +0.03740599, +0.02212356, +0.00318127],
+        ]
+
+        self.assertTrue(np.allclose(result, covar[:, :, 4, 4, 4, 4], atol=1e-4))
```

Comment:
I did see the `atol` is set to `1e-4` for in `test_covar3d.py`. This means that a lots of numbers in the matrix are smaller than this tolerance.  @janden should check it is acceptable or not.

Reply:
Sure, I'll talk to him soon.

fwiw, given the two ways to patch this, I picked the one that was consistent with the rest of the file(s), using the 1E-4 atol for the singles class it is running. The other way is fine too, just let me know, np.

This relates to other concerns about numerical precision/consistency in the python code. I'll make generalized issues tomorrow like we discussed. It is pretty straightforward to reconcile I think.

Reply:
The Mac and Linux version is OK to the default/smaller value. So very probably there is a bug. We need to do more evaluation. It might be OK to set a larger value  temporarily if we have create a new issue for this.

Reply:
Sure, it would be nice to narrow it down more .This particular diff didn't disturb me too much.  I suspect if we run on more varied systems we would encounter the same, not just on windows.  Just a hunch though...

Note, in the file, this was one of the few places that is without an adjusted tolerance.  I think it probably would be reconsidered in #149 . All in all, I think stabilizing the unit tests is more important than making them tighter atm, particularly because I plan to revisit testing/types soon-ish.  There are two or three testing problem I'd like to address first (this being one), before I move on.


```
tests/test_simulation.py:        self.assertTrue(np.allclose(blobs, np.load(os.path.join(DATA_DIR, 'sim_blobs.npy'))))
tests/test_simulation.py:        self.assertTrue(np.allclose(
tests/test_simulation.py:        self.assertTrue(np.allclose(images, np.load(os.path.join(DATA_DIR, 'sim_clean_images.npy')), rtol=1e-2))
tests/test_simulation.py:        self.assertTrue(np.allclose(images, np.load(os.path.join(DATA_DIR, 'sim_images_with_noise.npy')), rtol=1e-2))
tests/test_simulation.py:        self.assertTrue(np.allclose(images, np.load(os.path.join(DATA_DIR, 'sim_clean_images.npy')), rtol=1e-2))
tests/test_simulation.py:        self.assertTrue(images.shape, (8, 8, 25))
tests/test_simulation.py:        self.assertTrue(np.allclose(
tests/test_simulation.py:        self.assertTrue(
tests/test_simulation.py:        self.assertTrue(np.allclose([4.72837704, -4.72837709], coords, atol=1e-4))
tests/test_simulation.py:        self.assertTrue(np.allclose([8.20515764e-07, 1.17550184e-06], norms, atol=1e-4))
tests/test_simulation.py:        self.assertTrue(np.allclose([3.78030562e-06, -4.20475816e-06], inners, atol=1e-4))
tests/test_simulation.py:        self.assertTrue(
tests/test_simulation.py:        self.assertTrue(np.allclose(result['err'], 2.664116055950763, atol=1e-4))
tests/test_simulation.py:        self.assertTrue(np.allclose(result['rel_err'], 0.1765943704851626, atol=1e-4))
tests/test_simulation.py:        self.assertTrue(np.allclose(result['corr'], 0.9849211540734224, atol=1e-4))
tests/test_simulation.py:        self.assertTrue(np.allclose(result['err'], 13.322721549011165, atol=1e-4))
tests/test_simulation.py:        self.assertTrue(np.allclose(result['rel_err'], 0.5958936073938558, atol=1e-4))
tests/test_simulation.py:        self.assertTrue(np.allclose(result['corr'], 0.8405347287741631, atol=1e-4))
tests/test_simulation.py:        self.assertTrue(np.allclose(
tests/test_simulation.py:        self.assertTrue(np.allclose(
tests/test_simulation.py:        self.assertTrue(np.allclose(
```

Reply:
I agree this feels a little loose, but it's better to get the tests running and sort out these dtype and tolerance issues later. The goal should be to use the default tolerances unless there's a compelling reason not to, in which case that reason should be clearly stated in a comment for the unit test.

Reply:
Cool, yes I think we are all in agreement about potentially improving the tolerance.